src/mlpack/tests/distribution_test.cpp

```diff
@@ -444,6 +444,148 @@ BOOST_AUTO_TEST_CASE(GammaDistributionTrainTest)
 
 /**
  * This test verifies that the fitting procedure for GammaDistribution works
+ * properly when probabilities for each sample is given.
+ */
+BOOST_AUTO_TEST_CASE(GammaDistributionTrainWithProbabilitiesTest)
+{
+  double alphaReal = 5.4;
+  double betaReal = 6.7;
+
+  // Create a gamma distribution random generator
+  std::default_random_engine generator;
+  std::gamma_distribution<double> dist(alphaReal, betaReal);
+
+  size_t N = 50000;
+  size_t d = 2;
+  arma::mat rdata(d, N);
+
+  for(size_t j = 0; j < d; j++)
+    for(size_t i = 0; i < N; i++)
+      rdata(j, i) = dist(generator);
+
+  // create a uniform distribution random generator
+  std::uniform_real_distribution<double> prob(0, 1);
+  arma::vec probabilities(N);
+
+  for(size_t i = 0; i < N; i++)
+    probabilities(i) = prob(generator);
+
+  // fit results with probabilities and data
+  GammaDistribution gDist;
+  gDist.Train(rdata, probabilities);
+
+  // fit results with only data
+  GammaDistribution gDist2;
+  gDist2.Train(rdata);
+
+  BOOST_REQUIRE_CLOSE(gDist2.Alpha(0), gDist.Alpha(0), 1);
+  BOOST_REQUIRE_CLOSE(gDist2.Beta(0), gDist.Beta(0), 1);
+
+  BOOST_REQUIRE_CLOSE(gDist2.Alpha(1), gDist.Alpha(1), 1);
+  BOOST_REQUIRE_CLOSE(gDist2.Beta(1), gDist.Beta(1), 1);
+
+  BOOST_REQUIRE_CLOSE(alphaReal, gDist.Alpha(0), 1);
+  BOOST_REQUIRE_CLOSE(betaReal, gDist.Beta(0), 1);
+
+  BOOST_REQUIRE_CLOSE(alphaReal, gDist.Alpha(1), 1);
+  BOOST_REQUIRE_CLOSE(betaReal, gDist.Beta(1), 1);
+}
+
+
+// This test ensures that the same result is obtained when
+// trained with probabilities all set to 1 and with no probabilities at all.
+BOOST_AUTO_TEST_CASE(GammaDistributionTrainALLProbabilities1Test)
+{
+  double alphaReal = 5.4;
+  double betaReal = 6.7;
+
+  // Create a gamma distribution random generator
+  std::default_random_engine generator;
+  std::gamma_distribution<double> dist(alphaReal, betaReal);
+
+  size_t N = 1000;
+  size_t d = 2;
+  arma::mat rdata(d, N);
+
+  for(size_t j = 0; j < d; j++)
+    for(size_t i = 0; i < N; i++)
+      rdata(j, i) = dist(generator);
+
+  // fit results with only data
+  GammaDistribution gDist;
+  gDist.Train(rdata);
+
+  // fit results with data and each probability as 1
+  GammaDistribution gDist2;
+  arma::vec allProbabilities1(N, arma::fill::ones);
+  gDist2.Train(rdata, allProbabilities1);
+
+  BOOST_REQUIRE_EQUAL(gDist2.Alpha(0), gDist.Alpha(0));
+  BOOST_REQUIRE_EQUAL(gDist2.Beta(0), gDist.Beta(0));
+
+  BOOST_REQUIRE_EQUAL(gDist2.Alpha(1), gDist.Alpha(1));
+  BOOST_REQUIRE_EQUAL(gDist2.Beta(1), gDist.Beta(1));
```

Comment:
We should make this `BOOST_REQUIRE_CLOSE()` with a tolerance of `1e-5` or so, because floating-point errors in some implementations could provide slightly different results.  It looks like your current implementation appears to produce the exact same result, but that may not be true in the future if the code is modified (but still valid), so we should be careful to try and make our test future-proof.